plugins/field-numpad/src/field_numpad.ts

```diff
@@ -0,0 +1,243 @@
+    super(value, min, max, precision, validator, config);
+  }
+
+  /**
+   * Constructs a FieldNumpad from a JSON arg object.
+   * @param options A JSON object with options (value, min, max, and
+   *                          precision).
+   * @return The new field instance.
+   * @package
+   * @nocollapse
+   */
+  static fromJson(options: FieldNumpadOptions): FieldNumpad {
+    return new FieldNumpad(
+        options.value, undefined, undefined, undefined, undefined, options);
+  }
+
+  /* eslint-disable @typescript-eslint/naming-convention */
+  /**
+   * Show the inline free-text editor on top of the text along with the numpad
+   *    editor.
+   * @param e Optional mouse event that triggered the field to
+   *     open, or undefined if triggered programmatically.
+   * @param quietInput Quiet input (prevent focusing on the editor).
+   */
+  protected showEditor_(e?: Event, quietInput?: boolean) {
+    // Always quiet the input for the super constructor, as we don't want to
+    // focus on the text field, and we don't want to display the modal
+    // editor on mobile devices.
+    super.showEditor_(e, true);
+
+    // Build the DOM.
+    const editor = this.dropdownCreate_();
+
+    Blockly.DropDownDiv.getContentDiv().appendChild(editor);
+
+    const sourceBlock = this.getSourceBlock() as Blockly.BlockSvg;
+
+    const primary = sourceBlock.getColour() || '';
+    const tertiary = sourceBlock.getColourTertiary() || '';
+    Blockly.DropDownDiv.setColour(primary, tertiary);
+
+    Blockly.DropDownDiv.showPositionedByField(
+        this, undefined);
+  }
+
+  /**
+   * Updates the numpad when the field rerenders.
+   */
+  protected render_() {
+    super.render_();
+  }
+
+  /**
+   * Creates a table cell in the provided table row
+   * @param name The text content of the cell.
+   * @param row The row the cell will be inserted.
+   * @param position The position in the row the cell will be inserted.
+   * @return The newly created cell.
+   */
+  static createCellInRow_(name: string, row: HTMLTableRowElement,
+      position: number): HTMLTableCellElement {
+    const cell = row.insertCell(position);
+    cell.className = 'blocklyMenuItem';
+    cell.textContent = name;
+    return cell;
+  }
+
+  /**
+   * Creates the numpad editor and adds event listeners.
+   * @return The newly created numpad.
+   */
+  private dropdownCreate_(): HTMLElement {
+    const wrapper = document.createElement('div') as HTMLElement;
+    wrapper.className = 'fieldNumpadContainer blocklyMenu blocklyNonSelectable';
+    const table = document.createElement('table');
+    table.className = 'fieldNumpad';
+
+    const firstRow = table.insertRow(0);
+    const key7 = FieldNumpad.createCellInRow_('7', firstRow, 0);
+    const key8 = FieldNumpad.createCellInRow_('8', firstRow, 1);
+    const key9 = FieldNumpad.createCellInRow_('9', firstRow, 2);
+    const keyDel = FieldNumpad.createCellInRow_('Del', firstRow, 3);
```

Comment:
'Del' and 'Clr' are user-visible strings.  They might need to be internationalized.  So extracting them into a property where they can be set would be great.  An argument might be made that the numbers should also be internationalizable (https://en.wikipedia.org/wiki/Arabic_numerals#Comparison_with_other_digits) but given that they will print western digits when pressed, that argument is less convincing.